### PR TITLE
add separate esm and commonjs build so we can support legacy commonjs apps

### PR DESCRIPTION
**Pull request by jacobheric**
This actually a pretty hairy operation. We have separate ts builds and configs for esm code, commonjs code and type generation. Additionally we have to monkey patch the protobuf-ts generated code to actually be esm module compliant. 

The crux of the setup is the main package.json exports section we publish to npm:

```
  "exports": {
    "./protos/*": {
      "types": "./types/protos/*.d.ts",
      "require": "./cjs/protos/*.js",
      "import": "./esm/protos/*.js",
      "default": "./esm/protos/*.js"
    }
  },
```

This directs clients referencing the code to the correct modules depending on the import style they use.

### CODE REVIEW

**Comment by dselans** [comment]
Is the generate time affected?

File: build/ts/generatePackages.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#/bin/sh
+
```
Comment (dselans) [inline build/ts/generatePackages.sh:2]: Same here btw - would stick in a `mkdir -p /node/esm && mkdir -p /node/cjs` to avoid something from blowing up.

**Review by dselans — APPROVED** [review]
One minor comment.

File: build/ts/generatePackages.sh (hunk 1 above)
Comment (jacobheric) [reply build/ts/generatePackages.sh:2]: 👍

**Pull request merged**